service/common/src/main/java/org/apache/polaris/service/context/DefaultRealmContextResolver.java

```diff
@@ -37,19 +37,24 @@ public DefaultRealmContextResolver(RealmContextConfiguration configuration) {
 
   @Override
   public RealmContext resolveRealmContext(
-      String requestURL, String method, String path, Map<String, String> headers) {
-
-    String realm;
+      String requestURL, String method, String path, Function<String, String> headers) {
+    String realm = resolveRealmIdentifier(headers);
+    return () -> realm;
+  }
 
-    if (headers.containsKey(configuration.headerName())) {
-      realm = headers.get(configuration.headerName());
+  private String resolveRealmIdentifier(Function<String, String> headers) {
+    String realm = headers.apply(configuration.headerName());
+    if (realm != null) {
       if (!configuration.realms().contains(realm)) {
-        throw new IllegalArgumentException("Unknown realm: " + realm);
+        throw new UnresolvableRealmContextException("Unknown realm: " + realm);
       }
```

Comment:
Personally, i would push the check down into the `RealmContextConfiguration` rather than exposing a set of valid realms and doing the check here. But for this change, I think the `UnresolvableRealmContextExcpetion` is the right change to make